File: 2. Queue/[662]二叉树最大宽度.java
//给定一个二叉树，编写一个函数来获取这个树的最大宽度。树的宽度是所有层中的最大宽度。这个二叉树与满二叉树（full binary tree）结构相同，但一些节
//点为空。 
//
// 每一层的宽度被定义为两个端点（该层最左和最右的非空节点，两端点间的null节点也计入长度）之间的长度。 
//
// 示例 1: 
//
// 
//输入: 
//
//           1
//         /   \
//        3     2
//       / \     \  
//      5   3     9 
//
//输出: 4
//解释: 最大值出现在树的第 3 层，宽度为 4 (5,3,null,9)。
// 
//
// 示例 2: 
//
// 
//输入: 
//
//          1
//         /  
//        3    
//       / \       
//      5   3     
//
//输出: 2
//解释: 最大值出现在树的第 3 层，宽度为 2 (5,3)。
// 
//
// 示例 3: 
//
// 
//输入: 
//
//          1
//         / \
//        3   2 
//       /        
//      5      
//
//输出: 2
//解释: 最大值出现在树的第 2 层，宽度为 2 (3,2)。
// 
//
// 示例 4: 
//
// 
//输入: 
//
//          1
//         / \
//        3   2
//       /     \  
//      5       9 
//     /         \
//    6           7
//输出: 8
//解释: 最大值出现在树的第 4 层，宽度为 8 (6,null,null,null,null,null,null,7)。
// 
//
// 注意: 答案在32位有符号整数的表示范围内。 
// Related Topics 树 
// 👍 227 👎 0


//leetcode submit region begin(Prohibit modification and deletion)
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    class Pair{
        private TreeNode node;
        private int id;

        public Pair(){}

        public Pair(TreeNode node, int id) {
            this.node = node;
            this.id = id;
        }
    }

    public int widthOfBinaryTree(TreeNode root) {
        Queue<Pair> Q = new LinkedList<>();
        if (root != null) {
            Q.offer(new Pair(root, 0));
        }
        int ans = 0;
        while (Q.size() > 0) {
            int qSize = Q.size();
            int start = -1;
            int end = -1;
            for (int i = 0; i < qSize; i++) {
                Pair curNode = Q.poll();
                if (start == -1) {
                    start = curNode.id;
                }
                end = curNode.id;
                if (curNode.node.left != null) {
                    Q.offer(new Pair(curNode.node.left, curNode.id<<1));
                }
                if (curNode.node.right != null) {
                    Q.offer(new Pair(curNode.node.right, (curNode.id<<1)+1));
                }
            }
            ans = Math.max(ans, end-start+1);
        }
        return ans;
    }
}
//leetcode submit region end(Prohibit modification and deletion)

File: 02. Queue/[662]二叉树最大宽度.py
# 给你一棵二叉树的根节点 root ，返回树的 最大宽度 。 
# 
#  树的 最大宽度 是所有层中最大的 宽度 。 
# 
#  
#  
#  每一层的 宽度 被定义为该层最左和最右的非空节点（即，两个端点）之间的长度。将这个二叉树视作与满二叉树结构相同，两端点间会出现一些延伸到这一层的 
# null 节点，这些 null 节点也计入长度。 
#  
#  
# 
#  题目数据保证答案将会在 32 位 带符号整数范围内。 
# 
#  
# 
#  示例 1： 
#  
#  
# 输入：root = [1,3,2,5,3,null,9]
# 输出：4
# 解释：最大宽度出现在树的第 3 层，宽度为 4 (5,3,null,9) 。
#  
# 
#  示例 2： 
#  
#  
# 输入：root = [1,3,2,5,null,null,9,6,null,7]
# 输出：7
# 解释：最大宽度出现在树的第 4 层，宽度为 7 (6,null,null,null,null,null,7) 。
#  
# 
#  示例 3： 
#  
#  
# 输入：root = [1,3,2,5]
# 输出：2
# 解释：最大宽度出现在树的第 2 层，宽度为 2 (3,2) 。
#  
# 
#  
# 
#  提示： 
# 
#  
#  树中节点的数目范围是 [1, 3000] 
#  -100 <= Node.val <= 100 
#  
# 
#  Related Topics 树 深度优先搜索 广度优先搜索 二叉树 👍 537 👎 0


# leetcode submit region begin(Prohibit modification and deletion)
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def widthOfBinaryTree(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        cur_level = list()
        ans = 0
        if root:
            cur_level.append((root, 0))

        while cur_level:
            next_level = list()
            start = -1
            end = 0
            for i in range(len(cur_level)):
                cur_node = cur_level[i]
                if start == -1:
                    start = cur_node[1]
                end = cur_node[1]
                if cur_node[0].left:
                    next_level.append((cur_node[0].left, 2 * cur_node[1]))
                if cur_node[0].right:
                    next_level.append((cur_node[0].right, 2 * cur_node[1] + 1))
            cur_level = next_level
            ans = max(ans, end - start + 1)
        return ans
# leetcode submit region end(Prohibit modification and deletion)
